zerver/actions/user_groups.py

```diff
@@ -442,18 +453,38 @@ def do_change_user_group_permission_setting(
     old_value = getattr(user_group, setting_name)
     setattr(user_group, setting_name, setting_value_group)
     user_group.save()
+
+    old_setting_api_value = get_group_setting_value_for_api(old_value)
+    new_setting_api_value = get_group_setting_value_for_api(setting_value_group)
+
+    if not hasattr(old_value, "named_user_group") and hasattr(
+        setting_value_group, "named_user_group"
+    ):
+        # We delete the UserGroup which the setting was set to
+        # previously if it does not have any linked NamedUserGroup
+        # object, as it is not used anywhere else. A new UserGroup
+        # object would be created if the setting is later set to
+        # a combination of users and groups.
+        old_value.delete()
+
     RealmAuditLog.objects.create(
         realm=user_group.realm,
         acting_user=acting_user,
         event_type=RealmAuditLog.USER_GROUP_GROUP_BASED_SETTING_CHANGED,
         event_time=timezone_now(),
         modified_user_group=user_group,
         extra_data={
-            RealmAuditLog.OLD_VALUE: old_value.id,
-            RealmAuditLog.NEW_VALUE: setting_value_group.id,
+            RealmAuditLog.OLD_VALUE: get_group_setting_value_for_audit_log_data(
+                old_setting_api_value
+            ),
+            RealmAuditLog.NEW_VALUE: get_group_setting_value_for_audit_log_data(
```

Comment:
This is because the `extra_data` field needs a JSON serializable object and thus we cannot set these to a dataclass. This worked previously because we used TypedDict.

Reply:
Makes sense! Probably worth having `get_group_setting_value_for_audit_log_data` include a comment explaining that detail.